fix: networktransform SwitchTransformSpaceWhenParented forces updates when InLocalSpace is true

NoelStephensUnity opened the pull request:
## Purpose of this PR
This PR resolves the issue where setting SwitchTransformSpaceWhenParented to true and then setting InLocalSpace to true will force the motion authority to continually send updates each tick (whether there is a delta or not).

### Jira ticket
[MTTB-1584](https://jira.unity3d.com/browse/MTTB-1584)

fix: #3653
fix: #3646
fix: #3645 

### Changelog
- Added: NetworkRigidbody documentation section.
- Fixed: NetworkTransform state synchronization issue when `NetworkTransform.SwitchTransformSpaceWhenParented` is enabled and the associated NetworkObject is parented multiple times in a single frame or within a couple of frames.
- Fixed: Issue when spawning, parenting, and immediately re-parenting.
- Changed: NetworkShow sends a message and forces a NetworkVariable synchronization at the end of the frame prior to generating the CreateObjectMessage as opposed to waiting until the next network tick to synchronize the show with the update to NetworkVariables.
- Changed: NetworkTransform now synchronizes `NetworkTransform.SwitchTransformSpaceWhenParented` when it is updated by the motion model authority.
- Changed: NetworkObjects pending to be shown to clients can now occur on partial network ticks. If any pending NetworkObjects pending to be shown to clients happens to be ready on a new network tick they still are shown after network variable deltas have been processed.
- Changed: The default `NetworkDelivery` used by all messages is now reliable fragmented sequenced with the exception of named, unnamed, and any messages sent with a user specified network delivery type. This assures certain order of operations to be preserved when same call-stack changes are applied to a newly spawned, authority side, NetworkObject.
- Changed: NetworkTransform documentation to better reflect the Teleport methods intended usage along with updates to NetworkObject and physics areas of the documentation.

## Documentation Included

- Includes documentation for existing public API entry points:
  - Updates to the SwitchTransformSpaceWhenParented XML API that provides additional information about usage.
  - Updates to the SetState and Teleport for more clarity on intended usage for both network topologies.
  - Updates to NetworkObject.SpawnWithOwnership to call out how spawning with ownership can impact things like NetworkTransform during the spawning sequence when using a client-server topology.
- Includes updates to existing documentation:
  - Updates to NetworkTransform regarding the use of SwitchTransformSpaceWhenParented.
    - Add section covering spawning, order of operations, and how timing and network topology can impact the end result.
      - Include SetState/Teleport, spawning (with or without) ownership, and the authority motion model.
  - Updates to the NetworkObject spawning section
    - Expand the coverage to include:
      - Spawning then changing ownership vs SpawnWithOwnership
      - Spawning, NetworkTransform, and parenting.
  - Updated physics
    - Includes parenting physics objects using attachables.
  - Added a first pass (still could use some additions in the future) NetworkRigidbody section (pulled this part out of physics).

  
## Testing & QA (How your changes can be verified during release Playtest)
- Includes an integration test to validate the fix (`NetworkTransformAutoParenting`)
- Includes several additional integration tests via `NetworkTransformOrderOfOperations` to validate other edge case scenarios:
  - Test-1:
    - Server-> Spawn, change ownership, (wait), parent
  - Test-2:
    - Server-> Spawn, change parent, change ownership
  - Test-3:
    - Server-> Spawn, change parent (1), change ownership
    - Client-> PostSpawn re-parent (2)
  - Test-4:
    - Server-> Spawn no observers, change parents (multiple), NetworkShow, change ownership, Teleport RPC
    - Client-> Teleport RPC
  - Test-5:
    - Server-> Spawn, change parent (1), change ownership, Teleport RPC with NetworkBehaviourReference
    - Client-> Teleport RPC using NetworkBehaviourReference
  - Test-6:
    - Server-> Spawn, change ownership, change parent (1), Teleport RPC with NetworkBehaviourReference
    - Client-> Teleport RPC using NetworkBehaviourReference
  - Test-7:
    - Server-> Spawn with ownership, change parent (1), Teleport RPC with NetworkBehaviourReference
    - Client-> Teleport RPC using NetworkBehaviourReference
  - Test 8:
    - Server-> Start coroutine:
      - Coroutine: Spawn with no observers, change parent (1), NetworkShow, change ownership, Teleport RPC with NetworkBehaviourReference
    - Client-> Teleport RPC using NetworkBehaviourReference.
  - Test 9:
    - Server-> Start coroutine:
      - Coroutine: Spawn with no observers, NetworkShow, change parent (1)
  - Test 10:
    - Server-> Spawn with ownership, change parent (1), Wait (1),  Parent RPC with NetworkObjectReference
    - Client-> Re-parent (2) RPC using NetworkObjectReference
  - Test 11:
    - Server-> Spawn with ownership, Parent RPC with NetworkObjectReference,
    - Client-> Parent (1) RPC using NetworkObjectReference
    - Server-> Parent changed --> parent under 2nd object (was a delay, but the recommended way will be to wait if you know the client is going to parent.
- Includes a manual test ([need to import this small test into our manual tests](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/3646#issuecomment-3276996559))


### Functional Testing
[//]: # (If checked, List manual tests that have been performed.)
_Manual testing :_
- [ X ] `Manual testing done`
  - The original [CharacterControllerMovingBodies](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/tree/develop-2.0.0/Examples/CharacterControllerMovingBodies) project.
    - This verifies the existing functionality was not broken.
  - The manual test mentioned above that was used to replicate the user's issues and fix them.
    - This verifies (n) parenting events can occur on a single frame and it will be synchronized.

_Automated tests:_
- [ ] `Covered by existing automated tests`
- [ X ] `Covered by new automated tests`
  - (wip)

_Does the change require QA team to:_

- [ ] `Review automated tests`?
- [ ] `Execute manual tests`?
- [ ] `Provide feedback about the PR`?

If any boxes above are checked the QA team will be automatically added as a PR reviewer.

## Backports

This is an NGO v2.x.x specific issue. No backport is required.

NoelStephensUnity committed:
fix

Assure that the local InLocalSpace field is updated to match the parented status.

NoelStephensUnity committed:
fix

Reset IsTeleporting and ExplicitSet after invoking OnAuthorityPushTransformState so these values are preserved for user script.

Automatically adjust InLocalSpace when spawning a NeworkObject based on the NetworkObject's current parented status when SwitchTransformSpaceWhenParented is enabled.

NoelStephensUnity committed:
update

Invoking the OnAuthorityPushTransformState and OnNetworkTransformStateUpdated when synchronizing in order to provide users with the NetworkTransformState used to synchronize a NetworkTransform when first spawned.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Revert moving where IsTeleportingNextFrame and ExplicitSet are reset.
Add WasTeleported to NetworkTransformState in order to preserve whether the pushed state was teleported or not.

NoelStephensUnity committed:
Merge branch 'fix/networktransform-autoswitchspace-forcing-updates-localspace' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Adding documentation updates.

NoelStephensUnity committed:
update

XML API update

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
````diff
@@ -46,6 +46,8 @@ To spawn NetworkObjects with ownership use the following:
 GetComponent<NetworkObject>().SpawnWithOwnership(clientId);
 ```
+> [!NOTE]
+> When using `SpawnWithOwnership` method in a client-server network topology you must be aware that during the spawn sequence any component that has owner specific checks to perform specific actions will not be invoked on the server side. This can impact things like [NetworkTransform](../helper/networktransform.md) when set to an owner authority motion model. To avoid issues pertaining to this it ...
````
```suggestion
> When using the `SpawnWithOwnership` method in a client-server network topology, be aware that any component that has owner-specific checks to perform specific actions will not be invoked on the server side during the spawn sequence. This can impact things like [NetworkTransform](../helper/networktransform.md) when set to an owner authority motion model. To avoid issues, it's recommended to use `Spawn` where the server is the owner during the spawn sequence and then immediately follow with a call to `ChangeOwnership`. These two actions will get combined into a single `CreateObjectMessage` and avoid potential latency-driven issues.
```

jabbacakes commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
````diff
@@ -58,6 +60,9 @@ To give ownership back to the server use the `RemoveOwnership` method:
 GetComponent<NetworkObject>().RemoveOwnership();
 ```
+> [!NOTE]
+> It is not recommended to use `RemoveOwnership` when using a distributed authority network topology.
````
```suggestion
> Using `RemoveOwnership` in a distributed authority network topology isn't recommended.
```
This might be obvious from context, but why isn't it recommended?

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -1434,6 +1444,10 @@ internal bool SynchronizeScale
         /// authoritative side. Non-authoritative instances will be overridden by the next
         /// authoritative state update.
+        ///
+        /// !! Note !! <br />
```
```suggestion
```

NoelStephensUnity replied:
Resolved in the more recent commit.

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -1434,6 +1444,10 @@ internal bool SynchronizeScale
+        ///
+        /// !! Note !! <br />
+        /// When <see cref="SwitchTransformSpaceWhenParented"/> is enabled, this field will be automatically
+        /// adjusted. It is not recommended adjusting this field during runtime and when <see cref="SwitchTransformSpaceWhenParented"/> is enabled.
```
```suggestion
        /// adjusted. Adjusting this field during runtime and when <see cref="SwitchTransformSpaceWhenParented"/> is enabled isn't recommended.
```

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -1447,7 +1461,9 @@ internal bool SynchronizeScale
         /// Only works with <see cref="NetworkTransform"/> components that are not paired with a <see cref="NetworkRigidbody"/> or <see cref="NetworkRigidbody2D"/> component that is configured to use the rigid body for motion.<br />
         /// <see cref="TickSyncChildren"/> will automatically be set when this is enabled.
-        /// Does not auto-synchronize clients if changed on the authority instance during runtime (i.e. apply this setting in-editor).
+        /// This field is not auto-synchronize with non-authority clients if changed on the authority instance during runtime (i.e. apply this setting in-editor).
```
```suggestion
        /// This field doesn't auto-synchronize with non-authority clients if changed on the authority instance during runtime (so you should apply this setting in-Editor).
        /// Read the NetworkTransform documentation for more information and to avoid improper use.
```

NoelStephensUnity replied:
I changed this functionality and need to update the documentation again.  😄

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -1447,7 +1461,9 @@ internal bool SynchronizeScale
-        /// Does not auto-synchronize clients if changed on the authority instance during runtime (i.e. apply this setting in-editor).
+        /// This field is not auto-synchronize with non-authority clients if changed on the authority instance during runtime (i.e. apply this setting in-editor).
+        /// !! Note !! <br />
+        /// It is highly recommended to read the companion NetworkTransform documentation regarding this feature to avoid improper usage.
```
```suggestion
```

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -3876,7 +3902,9 @@ internal override void InternalOnNetworkObjectParentChanged(NetworkObject parent
-        ///  just the desired one(s)
+        ///  just the desired one(s).
+        /// !! Note !! <br />
+        /// It is not recommended to use this method during the spawn sequence. Refer to the NetworkTransform documentation for more information on the recommended usage.
```
```suggestion
        /// Using this method during the spawn sequence isn't recommended. Refer to the NetworkTransform documentation for more information on the recommended usage.
```

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -4013,10 +4041,12 @@ private void SetStateServerRpc(Vector3 pos, Quaternion rot, Vector3 scale, bool
         /// <summary>
         /// Teleport an already spawned object to the given values without interpolating.
+        /// !! Note !! <br />
+        /// It is not recommended to use this method during the spawn sequence. Refer to the NetworkTransform documentation for more information on the recommended usage.
```
```suggestion
        /// Using this method during the spawn sequence isn't recommended. Refer to the NetworkTransform documentation for more information on the recommended usage.
```

jabbacakes commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1941,6 +1941,12 @@ public void Spawn(bool destroyWithScene = false)
         /// </summary>
+        /// <remarks>
+        /// When using a client-server or distributed authority network topology, you should take into consideration any components
+        /// that might require ownership checks while running through the spawn process. To avoid issues that could arise by initialy spawning
```
```suggestion
        /// that might require ownership checks while running through the spawn process. To avoid issues that could arise by initially spawning
```

jabbacakes committed:
Doc pass on NetworkTransform additions

jabbacakes reviewed (approved):
Left suggestions on the smaller changes - have pushed a commit with changes to the larger new additions to the NetworkTransform page, mostly just trying to improve clarity and reduce verbiage.

This page is getting very long, might break it up at some point, but will do that in a separate PR.

NoelStephensUnity committed:
fix

NetworkSpawnManager:
Relative to spawning only, process deferred messages after post spawn.

NetworkTransform:
Synchronize the SwitchTransformSpaceWhenParented flag when it changes on the authority side.
Perform an early check in CheckForStateChange for changes to SwitchTransformSpaceWhenParented.
Apply changes to SwitchTransformSpaceWhenParented when processing a state update on non-authority instance.

NoelStephensUnity committed:
Merge branch 'fix/networktransform-autoswitchspace-forcing-updates-localspace' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
refactor

A complete refactoring of how parenting is handled when NetworkTransform.SwitchTransformSpaceWhenParented is enabled to provide a complete 1:1 match of (n) back-to-back parenting actions and/or many parenting actions that occur over several frames. This preserves the order of operations, sends 1 full state update plus an added parenting directive per action immediately, and then upon receiving the non-authority instances will apply the parenting, transform the local values and interpolators' queued entries  between transform spaces (world to local, local to world, or local to local).

NoelStephensUnity commented:
/ci ngo

NoelStephensUnity committed:
style

Standards/style fix.

NoelStephensUnity commented:
/ci ngo

NoelStephensUnity committed:
style - PVP

More hidden white spaces...

NoelStephensUnity commented:
/ci ngo

NoelStephensUnity committed:
update

Removing one line space

NoelStephensUnity commented:
/ci ngo

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Adding change log entry.

NoelStephensUnity committed:
update

Check if writing as binary resolves the CRLF issue.

NoelStephensUnity committed:
update

Reverting last change.
Removing a single line of code.

NoelStephensUnity committed:
update

Migrating all NetworkTransform serialization into the NetworkTransformMessage.
Removing legacy observers for the NetworkTransform.

NoelStephensUnity committed:
upate

Increment count after check when iterating through observers.

NoelStephensUnity committed:
style

remove whitespaces.

NoelStephensUnity committed:
test

Add a test to validate this PR.

NoelStephensUnity committed:
Apply suggestions from code review

Co-authored-by: Amy Reeve <amy.reeve@unity3d.com>

NoelStephensUnity committed:
doc: update

Adjusting based on the PR's recent updates.
Refactoring the physics section.
Moved the NetworkRigidbody content out of physics and into a new NetworkRigidbody section under componetns.
Moved physics up to the components layer.
Did a large update to the physics documentation and provided an additional walk through of  how to "parent" physics bodies using attachables.

NoelStephensUnity committed:
doc: style

Adding one suggested change I left out.

NoelStephensUnity committed:
style - PVP

Fixing trailing spaces.

NoelStephensUnity committed:
update

Further clarifying the update to NetworkObject.SpawnWithOwnership

NoelStephensUnity committed:
fix

If half float precision is enabled, then update the current half float position state after transforming the current position due to a change in the NetworkObject's status.

NoelStephensUnity committed:
test and style

Updating the SwitchTransformSpaceWhenParented test to validate using various NetworkTransform configurations (i.e. half precision, quaternion, quaternion compressed) while also parenting with world position stays enabled and disabled.

Removing trailing space in recent documentation update.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

adding additional changelog entries.

NoelStephensUnity committed:
update

reflecting the current version of this file from main branch.

NoelStephensUnity committed:
update

Try this one more time.

NoelStephensUnity commented on Tools/scripts/Utils/general_utils.py:
```diff
@@ -1,4 +1,4 @@
-﻿"""Helper class for common operations."""
+"""Helper class for common operations."""
```
This is an artifact from earlier changes... just reflecting what it is currently in the main v2 branch.

NoelStephensUnity committed:
update

Making NetworkTransformState.SwitchTransformSpaceWhenParented internal as it doesn't really need to be public for its use case.

jabbacakes committed:
Docs pass on new content

jabbacakes committed:
Reinstating note

jabbacakes reviewed (approved):
Alrighty, that was quite a lot 😅  I've just been through and given it a bit of polish. Main change to note is that I've modified the walkthrough of AttachableBehaviour parenting to be a bit more docsy (fewer asides and less conversational tone).

jabbacakes committed:
Updating landing page

NoelStephensUnity committed:
refactor

Relatively large change:

- Preserving order of operations by making the default network delivery be reliable fragmented sequenced with the exception of named, unnamed, and any message types adjusted explicitly by user code to use a different network delivery method.

- Adjusted parenting directive to allow the server to force a parenting directive on a NetworkTransform it does not have the motion authority over.

- The NetworkObjects to show now will be also processed in-between ticks to preserve order of operations.
  - NetworkObjects pending to be shown on the same frame as the tick update will still be processed during the tick update.

NoelStephensUnity committed:
Merge branch 'fix/networktransform-autoswitchspace-forcing-updates-localspace' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Left out the half precision adjustment in the server-side edge case for back-to-back changes.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Adding additional change log entries.

NoelStephensUnity committed:
update

Minor version increment.

NoelStephensUnity committed:
fix

Fix the issue where a user can still set InLocalSpace when SwitchTransformSpaceWhenParented is enabled.

NoelStephensUnity committed:
fix

Fix issue where client owner authority NetworkTransform would not set the right IsParented status when sending a parenting directive.

NoelStephensUnity committed:
fix

Assure MessageDelivery is initialized whether in editor or stand alone.

NoelStephensUnity committed:
refactor

Include any pending dirty NetworkVariable updates just prior to sending the Show/CreateObjectMessage.

NoelStephensUnity committed:
style - PVP

Removing un-used comment line to avoid PVP-124-2

NoelStephensUnity committed:
refactor

Better organizing NetworkBehaviourUpdater's methods to be a bit more modular in order to better handle how we handle force synchronizing NetworkVariables when showing a NetworkObject.

Mobed all of that logic from NetworkBehaviourUpdater to the late update section in NetworkManager.

NoelStephensUnity committed:
fix

Needed to check the inverse of whether either are true to determine if we should exit early in the NetworkBehaviourUpdater.

NoelStephensUnity committed:
fix

Removing debug log entry that is was added for troubleshooting an issue while making changes in this PR.
Not throwing an exception when getting a message type by actual message type (only used in DA codec tests). Instead, it now returns the default reliable fragmented sequenced.

NoelStephensUnity committed:
fix

Try this MessageDelivery fix for the codec test one more time.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
refactor and fix

Removed the extra byte per updated from the parenting directive and added it as a flag.
Fixed spawn no observers, show, and then parent issue where the initial parenting during spawn was causing issues with a back-to-back parenting message.

NoelStephensUnity committed:
fix

Only invoke OnDestroy if Singleton is not null during the application quit.
Resolved the issue where a server spawning an object with ownership that then changes the parent in the same frame with a NetworkTransform that has SwitchTransformSpaceWhenParented enabled was not applying the parenting directive (being sent to all observers) locally.

NoelStephensUnity committed:
fix

Regressed a previous fix when resolving the most current issue where generating a parent directive should only apply the parenting directive if it is the server and not the motion authority (whether local or world) and to assure if we are the motion authority we still set the m_PreviousParent in the event ownership changes and the NetworkTransform is using an owner authority motion model.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Minor adjustment after merge.
Don't invoke NetworkManager.OnDestroy unless the singleton is null and the current instance is not listening.

NoelStephensUnity committed:
fix

If we suddenly stopped and trying to update non-existent NetworkTransform instances, then just reset the NetworkTransformTickRegistration system.

NoelStephensUnity committed:
fix

Allow NetworkObjects to run through the despawn process if they are destroyed by something like the UnityEngine SceneManager when unloading a scene. Added validation check against the NetworkObject's GameObject's scene (IsValid) and whether the scene is loaded or not to determine if an error message should be logged or if it is just being destroyed due to the scene being unloaded (abrupt unloading like exiting play mode while in a session).

NoelStephensUnity committed:
fix

If a NetworkTransformMessage is received while in the middle of shutting down, then do not attempt to process it.

NoelStephensUnity committed:
fix

Changing #3687 adjustment to use the version of SpawnNetworkObjectLocally that accepts a SceneObject which doesn't invoke pre-spawn but does invoke post spawn and processes deferred messages.
These were moved in this PR to assure post spawn was invoked after the object was 100% done locally spawning and as a last step process any deferred messages targeting this object.

NoelStephensUnity committed:
test

Made a few adjustments to this test in order to debug, added additional logging, and marked this test as needing to be updated (and extended).

NoelStephensUnity committed:
style

removing whitespaces

NoelStephensUnity committed:
test - fix

Validating clients had initialized was some how was encapsulated within a verbose debug check. Not sure why this was passing before other than certain messages might have always been processed out of order before...?

NoelStephensUnity committed:
fix and update

Removing the previously removed post spawn invocation from the common spawn method (missed when merging).

Renaming the local spawn methods to reflect the context of the local spawn (i.e. authority is the first to spawn and then notify where non-authority is invoked by the authority's CreateObjectMessage). It will help keep the context clear (possible area to merge into one with a flag to determine if it is an authority or non-authority spawn action).

NoelStephensUnity committed:
test

Minor adjustments to the test that caught this issue.
Added logs and condensed NetworkUpdate.

NoelStephensUnity committed:
style

insert a whitespace

NoelStephensUnity committed:
style - PVP

Removing whitespace at the end of the comment

NoelStephensUnity committed:
refactor

Minor performance gain by using a pre-defined static network delivery that is registered by message type  within MessageDelivery.

NoelStephensUnity committed:
fix

Removing using directive.
Reverting changes to NetworkVariableCollectionsTests

NoelStephensUnity committed:
update

Minor code clean up with a handful of additional comments.

NoelStephensUnity committed:
style

Adjusting internal comments for clarity purposes.

NoelStephensUnity committed:
refactor

This includes a different approach to handling  buffered linear interpolation conversion when switching transform space.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
update

Making parent internal.

NoelStephensUnity committed:
update

Removing additional script not used.

NoelStephensUnity committed:
refactor

Noticed an issue with the world space approach and rotating platforms (parents).
This is a refactor of the same kind of approach (i.e. interpolator handles transform space updates) with the addition of tracking each measurement's parent, handling any transform space conversion if the current parent is different from the measurement's parent, and just keeping everything transform space relative.

NoelStephensUnity committed:
fix

Set the target parent when initializing.
Added additional checks when getting the current interpolated value to handle the interpolation frames between the current and next measurement states.

NoelStephensUnity committed:
fix

This resolves the issue where there was a de-synch between converting transform spaces.

NoelStephensUnity committed:
update

Removing left over debug script.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
fix

This resolves the broken half float issue.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
fix

Fixing issue with synch when SwitchTransformSpaceWhenParented is disabled.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
fix

Allow a full teleport when using SwitchTransformSpaceWhenParented.

NoelStephensUnity committed:
fix - refactor

This refactors the previous "fix" to preserve the teleport flag but also keeps the original logic which is needed for when SwitchTransformSpaceWhenParented is disabled.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
test - wip

This includes a more modular framework for generating order of operation sequences in order to create a wide range of tests that validate specific order of operation actions are processed and applied on the non-authority instances as they were on the authority instance.

NoelStephensUnity committed:
test

Updated the order of operations test to include 10 of the 11 spawn sequence ordering tests.
Added `NetcodeIntegrationTest.SpawnObjectInstance` that allows you to create an instance of a prefab and pass the instance in to be spawned through the normal integration test spawn process.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
style

Removing single white space after an (internal) comment...

NoelStephensUnity committed:
style

removing more whitespace.

NoelStephensUnity committed:
Test-update

Some general clean up and separating areas of the NetworkTransformOrderOfOperations by region to help with navigating through the different parts.

NoelStephensUnity committed:
test

Adding the 11th test to the order of operations test.

NoelStephensUnity committed:
update

Adding missing change log entry.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
test

Check on the receiving side of the `ReferenceTeleportRpc` call that only the motion authority applies the teleport.

NoelStephensUnity committed:
test-fix

Moving the motion authority check within the check to assure the SpawnSequenceController NetworkBehaviourReference is valid.

NoelStephensUnity committed:
doc

Updated some XML API documentation and corrected a typo in a comment.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
fix

Merge issue.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

EmandM commented on com.unity.netcode.gameobjects/Documentation~/components/helper/networkrigidbody.md:
```diff
@@ -0,0 +1,42 @@
+# NetworkRigidbody
```
We should update the `NetworkRigidbody` help URL to point to this page

EmandM commented on com.unity.netcode.gameobjects/Documentation~/TableOfContents.md:
```diff
@@ -32,8 +32,12 @@
     * [NetworkTransform](components/helper/networktransform.md)
-      * [Physics](advanced-topics/physics.md)
+  * [Physics](advanced-topics/physics.md)
+* [Ownership and authority](ownership-authority.md)
```
I think this is a merge conflict with my ownership changes. I moved these Ownership pages and so I think these ownership pages are elsewhere now.

NoelStephensUnity replied:
Nice catch!

EmandM commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
I think this paragraph changed as well. There is a new ownership overview page that this section should link to.

EmandM commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
I would maybe ensure that the new ownership overview page is linked here. That way we don't need to try explain ownership in the note underneath.

```suggestion
To spawn a `NetworkObject` that is [owned](../../terms-concepts/ownership.md) by a different game client than the one doing the spawning, use the following:
```

EmandM commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
````diff
@@ -46,6 +46,9 @@ To spawn NetworkObjects with ownership use the following:
 GetComponent<NetworkObject>().SpawnWithOwnership(clientId);
 ```
+> [!NOTE]
+> When using the `SpawnWithOwnership` method, be aware that any component that has owner-specific checks to perform specific actions won't be invoked on the spawn authority  side during the spawn sequence. The spawn authority is the server when using a client-server network topology, and can be any client when using a distributed authority network topology. Using `SpawnWithOwnership` can impact t ...
````
If we link the ownership overview above, I think we can simplify this wording some.

```suggestion
> The `SpawnWithOwnership` method allows a game client to spawn an object that it doesn't own. This means any owner-specific checks during the spawn sequence will not be invoked on the spawn authority side. Using `SpawnWithOwnership` can result in unexpected behavior if the spawn authority makes any additional post-spawn adjustments within the same frame.
```

(though looking at what's in that ownership page, I should expand the info about the motion model authority on that page)

EmandM commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
````diff
@@ -46,6 +46,9 @@ To spawn NetworkObjects with ownership use the following:
 ```
+> [!NOTE]
+> When using the `SpawnWithOwnership` method, be aware that any component that has owner-specific checks to perform specific actions won't be invoked on the spawn authority  side during the spawn sequence. The spawn authority is the server when using a client-server network topology, and can be any client when using a distributed authority network topology. Using `SpawnWithOwnership` can impact t ...
+> To avoid potential issues, it's recommended to use `Spawn`, where the spawn authority starts as the owner throughout the spawn sequence, makes adjustments post-spawn, and then immediately follow with a call to `ChangeOwnership`.
````
Further re-wording

```suggestion
> To avoid potential issues, it's recommended to use `Spawn` if the spawn authority needs to make any adjustments post-spawn. After adjusting, the spawn authority can immediately follow with a call to `ChangeOwnership`.
```

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviourUpdater.cs:
```diff
@@ -10,23 +11,185 @@ public class NetworkBehaviourUpdater
         private NetworkManager m_NetworkManager;
         private NetworkConnectionManager m_ConnectionManager;
+
+        /// <summary>
```
🎉

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviourUpdater.cs:
```diff
@@ -10,23 +11,185 @@ public class NetworkBehaviourUpdater
+        {
+            for (int i = 0; i < m_ConnectionManager.ConnectedClientsList.Count; i++)
+            {
+                var client = m_ConnectionManager.ConnectedClientsList[i];
```
small nit:

```suggestion
            foreach (var client in m_ConnectionManager.ConnectedClientsList)
            {
```

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviourUpdater.cs:
```diff
@@ -10,23 +11,185 @@ public class NetworkBehaviourUpdater
+        /// Clients handle processing dirty objects relative to the client.
+        /// The <see cref="NetworkVariableDeltaMessage"/> is client to server.
+        /// With distributed authority live service sessions, this is sent to
+        /// the Rust server.
```
nit: we kind of avoid directly mentioning the language of the CMB service in the codebase

```suggestion
        /// the CMB service.
```

p.s I love this change! ❤️

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviourUpdater.cs:
```diff
@@ -10,23 +11,185 @@ public class NetworkBehaviourUpdater
+            if (!(m_NetworkManager.IsServer || networkObject.IsOwner))
+            {
+                return;
+            }
```
Just a question for DAHost/Client-server when the server isn't the owner: Does the server always want to do the following logic even when it doesn't own the `NetworkVariable`?

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1738,18 +1738,28 @@ private void OnDestroy()
+                    if (gameObject != null && gameObject.scene.IsValid() && gameObject.scene.isLoaded)
                     {
-                        if (NetworkManager.DistributedAuthorityMode)
+                        if (NetworkManager.LogLevel <= LogLevel.Error && gameObject != null && gameObject.scene.IsValid() && gameObject.scene.isLoaded)
```
nit: Is this if check duplicated?

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -2309,9 +2325,10 @@ private void OnTransformParentChanged()
             }
             var removeParent = false;
             var parentTransform = transform.parent;
+            var parentObject = (NetworkObject)null;
```
Out of my interest purely because of curiosity and maybe consistency; do you have a general preference between these two options?

```csharp
// var with typecast
var someVariable = (VariableType)null;

// direct typing
VariableType someVariable = null;
```

I tend to do the second, but I really very much do not have a preference. It'd be great to get your thoughts on it.

NoelStephensUnity committed:
Merge branch 'develop-2.0.0' into fix/networktransform-autoswitchspace-forcing-updates-localspace

EmandM commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ClientConnectedMessage.cs:
```diff
@@ -49,7 +49,7 @@ public void Handle(ref NetworkContext context)
             }
 
-            // DANGO-TODO: Remove the session owner object distribution check once the service handles object distribution
+            // This handles object redistribution when scene management is disabled
```
Any chance you know the status on the rust implementation of this?

EmandM commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/NetworkTransformMessage.cs:
```diff
@@ -114,22 +141,22 @@ public bool Deserialize(FastBufferReader reader, ref NetworkContext context, int
-                        var targetId = (ulong)0;
-                        for (int i = 0; i < targetCount; i++)
+                        var count = 0;
+                        foreach (var targetId in networkObject.Observers)
```
nit: Might be worth adding a comment here for future us about why we are using a `foreach` loop here rather than a `for` loop.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -1030,7 +1031,7 @@ internal NetworkObject CreateLocalNetworkObject(NetworkObject.SceneObject sceneO
         /// DAHost client and standard DA clients invoke this method.
         /// </summary>
-        internal void SpawnNetworkObjectLocally(NetworkObject networkObject, ulong networkId, bool sceneObject, bool playerObject, ulong ownerClientId, bool destroyWithScene)
+        internal void AuthorityLocalSpawn(NetworkObject networkObject, ulong networkId, bool sceneObject, bool playerObject, ulong ownerClientId, bool destroyWithScene)
```
Just an idea, we can add a `[NotNull]` compiler check on the `NetworkObject` parameter and save on the null check

```suggestion
        internal void AuthorityLocalSpawn([NotNull] NetworkObject networkObject, ulong networkId, bool sceneObject, bool playerObject, ulong ownerClientId, bool destroyWithScene)
```

EmandM commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -1095,6 +1096,41 @@ internal void SpawnNetworkObjectLocally(NetworkObject networkObject, ulong netwo
+            if (networkObject == null)
+            {
+                throw new ArgumentNullException(nameof(networkObject), "Cannot spawn null object");
+            }
```
Same idea for the compiler check here:

```suggestion
        internal void NonAuthorityLocalSpawn([NotNull] NetworkObject networkObject, in NetworkObject.SceneObject sceneObject, bool destroyWithScene)
        {
```

EmandM commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformAutoParenting.cs:
```diff
@@ -0,0 +1,330 @@
+            return !hadError;
+        }
+
+        private bool AllClientsSpawnedObject()
```
This is now a helper in `NetcodeIntegrationTest`! `WaitForSpawnedOnAllOrTimeOut` 😄

EmandM commented on com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md:
````diff
@@ -58,6 +61,9 @@ To give ownership back to the server use the `RemoveOwnership` method:
 GetComponent<NetworkObject>().RemoveOwnership();
 ```
+> [!NOTE]
+> Using `RemoveOwnership` in a distributed authority network topology isn't recommended.
````
We actually throw an error in `RemoveOwnership` with DA ([NetworkSpawnManager.cs:L427](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/d8f96d65ab90471893226766401b46059dd054f6/com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs#L427))

```suggestion
> `RemoveOwnership` isn't supported when using a [distributed authority network topology](../../terms-concepts/distributed-authority.md).
```

EmandM commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformOrderOfOperations.cs:
```diff
@@ -0,0 +1,1106 @@
+        private void ConfigureSequencesTest7(NetworkObject parent1)
+        {
+            SpawnSequenceController.CurrentTest = "Test7 (Client-Server Only)";
+            if (m_AuthorityNetworkManager.DistributedAuthorityMode)
```
I would rather if this check was in the test itself and this function was simply not called when in `DistributedAuthorityMode`. I think it adds too much complexity to "hide" the skip in here and inside `ShouldRun`.

e.g
```csharp
ConfigureSequencesTest5(m_AuthorityGenericInstances[0]);
yield return RunTestSequences(spawnWithObservers: false);

if (!m_DistributedAuthority) // or m_NetworkTopologyType == NetworkTopologyTypes.ClientServer
{
    ConfigureSequencesTest6(m_AuthorityGenericInstances[0]);
    yield return RunTestSequences();
    
    ConfigureSequencesTest7(m_AuthorityGenericInstances[0]);
    yield return RunTestSequences(spawnWithOwnership: true);
}

ConfigureSequencesTest8(m_AuthorityGenericInstances[0]);
yield return RunTestSequences(spawnWithObservers: false);

```

EmandM commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformOrderOfOperations.cs:
```diff
@@ -0,0 +1,1106 @@
+            m_AuthorityNetworkManager = GetAuthorityNetworkManager();
+            yield return SpawnGenericParents();
+
+            ConfigureSequencesTest1(m_AuthorityGenericInstances[0]);
```
Do you think we could put `m_AuthorityGenericInstances[0]` and `m_AuthorityGenericInstances[1]` into variables with clearer names?

I think the xml doc is ok enough for the `ConfigureSequencesTest` naming pattern.

EmandM commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformOrderOfOperations.cs:
```diff
@@ -0,0 +1,1106 @@
+                    }
+                }
+
+                // Assure all sequenced actions have been invoked.
+                yield return WaitForConditionOrTimeOut(SpawnSequenceController.AllActionsInvoked);
+                if (s_GlobalTimeoutHelper.HasTimedOut())
+                {
+                    // If we timed out, then check for pending and if found wait for the condition
+                    // once more.
+                    if (SpawnSequenceController.ActionIsPending())
+                    {
+                        yield return WaitForConditionOrTimeOut(SpawnSequenceController.AllActionsInvoked);
+                    }
+                }
+                AssertOnTimeout($"[{SpawnSequenceController.CurrentTest}] Not all actions were invoked for the current test sequence!\n {SpawnSequenceController.ErrorLog}");
```
nit: My personal preference is to use a local timeout helper here rather than the global one. Just avoids using "magic" state and makes it clearer what's being checked where.

```suggestion
                // Assure all sequenced actions have been invoked.
                var helper = new TimeoutHelper();
                yield return WaitForConditionOrTimeOut(SpawnSequenceController.AllActionsInvoked, helper);
                if (helper.HasTimedOut())
                {
                    // If we timed out, then check for pending
                    if (SpawnSequenceController.ActionIsPending())
                    {
                        // Restart the timeout and wait again for the pending actions
                        helper = new TimeoutHelper();
                        yield return WaitForConditionOrTimeOut(SpawnSequenceController.AllActionsInvoked, helper);
                    }
                }
                AssertOnTimeout($"[{SpawnSequenceController.CurrentTest}] Not all actions were invoked for the current test sequence!\n {SpawnSequenceController.ErrorLog}", helper);
```

EmandM commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformOrderOfOperations.cs:
```diff
@@ -0,0 +1,1106 @@
+                // Assure the generic parents are all at the root hierarchy.
+                foreach (var parent in m_AuthorityGenericInstances)
+                {
+                    if (parent.transform.parent != null)
+                    {
+                        parent.transform.parent = null;
+                    }
```
nit: we may as well set it to `null` directly. Makes the logic a bit simpler and saves the equals conversion seeing as we're going to set it anyway.

```suggestion
                    parent.transform.parent = null;
```

EmandM commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformOrderOfOperations.cs:
The mechanism in this test is super cool! I really like what you've done here. However, I am a little bit worried about ongoing maintenance.

Is there any chance you can add more xml doc to the top of this file to add an overview on how it works? It's definitely clean and understandable after spending time with it, but I'm worried about trying to understand it in a rush.

EmandM commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -10,12 +10,21 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Added
 
+- Added a `Set` function onto `NetworkList` that takes an optional parameter that forces an update to be processed even if the current value is equal to the previous value. (#3690)
```
Hmm, is there a merging issue here?

NoelStephensUnity replied:
Oh... yeah that does indeed seem to be one.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -1433,35 +1498,38 @@ internal bool SynchronizeScale
         /// Only works with <see cref="NetworkTransform"/> components that are not paired with a <see cref="NetworkRigidbody"/> or <see cref="NetworkRigidbody2D"/> component that is configured to use the rigid body for motion.<br />
         /// <see cref="TickSyncChildren"/> will automatically be set when this is enabled.
-        /// Does not auto-synchronize clients if changed on the authority instance during runtime (i.e. apply this setting in-editor).
+        /// This field is auto-synchronize with non-authority clients if changed by the authority instance.
```
tiny nit: 

```suggestion
        /// This field is auto-synchronized with non-authority clients when changed by the authority instance.
```

EmandM commented on com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs:
```diff
@@ -4468,10 +4573,11 @@ internal NetworkTransformState OutboundState
             {
                 // TODO: Investigate where this state should be applied or just discarded.
                 // For now, discard the state if we assumed ownership.
+                // Debug.Log($"[Client-{NetworkManager.LocalClientId}] Ignoring inbound update from Client-{0} and parentUpdated:{isParentingDirective}!");
```
Do we mean to leave this log commented out?

NoelStephensUnity replied:
Yep... this is an intentional "blemish" so we will eventually circle back to the issue... in-flight messages from a previous owner after ownership has already changed.

This requires putting together a manual test example that can get the in-flight messages from a previous owner to replicate consistently. Then we can determine how much of an impact the 1 or 2 latent messages might have in the over-all interpolation. This has a higher chance of happening when using a distributed authority network topology than a client-server.

EmandM replied:
I love the idea of an intentional blemish!

EmandM commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/NetworkTransformMessage.cs:
```diff
@@ -32,7 +32,28 @@ public void Serialize(FastBufferWriter writer, int targetVersion)
+                var position = writer.Position;
+                // Provides the source of the message (NetworkObject-->NetworkTransform : NetworkBehaviour).
+                BytePacker.WriteValueBitPacked(writer, NetworkTransform.NetworkObjectId);
+#if NGO_NETWORKTRANSFORMSTATE_LOGWRITESIZE
```
Do we want to document this scripting define somewhere? Or leave it just as internal info?

NoelStephensUnity replied:
For right now let's keep it internal... but it would be nice if we could add additional metrics like this to users.
Just requires us to come up with a well-defined strategy for this.

EmandM commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -453,6 +448,17 @@ public void NetworkUpdate(NetworkUpdateStage updateStage)
+                        // Handles object redistribution when scene management is disabled and
+                        // using a distributed authority network topology. Only set specific to
+                        // this configuration and when a client connects.
+                        if (RedistributeToClients)
```
Out of interest, do you know where this is being set?

NoelStephensUnity replied:
Within the [ClientConnectedMessage](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/bcb73cda29b5aaa33c5a4f56656fb9ead3b0d3fa/com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ClientConnectedMessage.cs#L74)

EmandM reviewed (commented):
This PR was genuinely a joy to review :godmode: 

So many nice things in here I couldn't begin to list them all. Most of my comments are small nit picks. Overall this looks so clean and nice and wonderful I don't even have enough words to say how much I like it.

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Components/NetworkTransform.cs

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Documentation~/components/core/networkobject.md

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
Apply suggestion from @EmandM

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
Apply suggestion from @EmandM

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
Apply suggestion from @EmandM

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
update

Minor merge fixes to table of contents.
Helper URLs now point to the rigidbody documentation.
Adding to be used best practices marker.

NoelStephensUnity committed:
doc - fix

Fixing some [!NOTE] sections where text was on the same line (which breaks the notation check when converting the documentation).

NoelStephensUnity committed:
PVP

Removing trailing white space

NoelStephensUnity committed:
fix

replaced .md with .html

NoelStephensUnity committed:
test - fix

HelpURL tests now check to see if the document exists before failing if it can't be loaded from the latest documentation update.

Adjusted the HelpUrls base portions of the URL so they are internal.

NoelStephensUnity committed:
test - refactor

Using System.IO.File.
Migrating into the testproject's runtime tests.
Encapsulating within UNITY_EDITOR define.
Removed the iPhone platform exclusion.

NoelStephensUnity committed:
test-fix

Fixing import issue.

NoelStephensUnity committed:
Apply suggestion from @EmandM

Co-authored-by: Emma <emma.mcmillan@unity3d.com>

NoelStephensUnity committed:
fix

Removing duplicate logic checks in despawn.

NoelStephensUnity committed:
style

Adding comment.

NoelStephensUnity committed:
refactor

Refactoring the NetworkObject check.

NoelStephensUnity committed:
update

Adjusting test and version bump

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviourUpdater.cs

Co-authored-by: Emma <emma.mcmillan@unity3d.com>
